File: Go/add_two_numbers/2.go
package addtwonumbers

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummyHead := ListNode{0, nil}
	tail := &dummyHead
	carry := 0

	getVal := func(l *ListNode) int {
		if l != nil {
			return l.Val
		}
		return 0
	}

	getNext := func(l *ListNode) *ListNode {
		if l != nil {
			return l.Next
		}
		return nil
	}

	for l1 != nil || l2 != nil || carry != 0 {
		digit1 := getVal(l1)
		digit2 := getVal(l2)
		l1 = getNext(l1)
		l2 = getNext(l2)

		sum := digit1 + digit2 + carry
		digit := sum % 10
		carry = sum / 10

		newNode := &ListNode{digit, nil}
		tail.Next = newNode
		tail = tail.Next

	}

	return dummyHead.Next
}

File: TypeScript/add_two_numbers/2.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function addTwoNumbers(
  l1: ListNode | null,
  l2: ListNode | null,
): ListNode | null {
  const dummyHead = new ListNode();
  let tail = dummyHead;
  let carry = 0;

  while (l1 != null || l2 != null || carry != 0) {
    const digit1 = l1?.val ?? 0;
    const digit2 = l2?.val ?? 0;
    l1 = l1?.next ?? null;
    l2 = l2?.next ?? null;

    const sum = digit1 + digit2 + carry;
    const lastDigit = sum % 10;
    carry = Math.floor(sum / 10);

    const newNode = new ListNode(lastDigit);
    tail.next = newNode;
    tail = tail.next;
  }

  return dummyHead.next;
}

export { addTwoNumbers, ListNode };
